rfcs/45-stageless.md

```diff
@@ -0,0 +1,727 @@
+For example, by default, each app stores both a startup and main schedule: the former runs only once on app creation, while the latter loops endlessly.
+
+The main and startup schedules can be accessed using the `CoreSchedule::Main` and `CoreSchedule::Startup` labels respectively.
+However, schedules are a surprisingly powerful and flexible tool: they can be used to handle initialization and cleanup logic when working with states or used ad-hoc to run game logic in complex patterns.
+By default, systems are added to the main schedule.
+You can control this by adding the `.in_set_schedule(ScheduleLabel::Variant)` system descriptor to your system or system set.
+
+#### Startup systems
+
+Startup systems are stored in their own schedule, with the `CoreSchedule::Startup` label.
+When using the runner added by `MinimalPlugins` and `DefaultPlugins`, this schedule will run exactly once on app startup.
+
+You can add startup systems with the `.add_startup_system(on_startup)` method on `App`, which is simply sugar for `.add_system(on_startup.in_set_schedule(CoreSchedule::Startup))`.
+
+### System configuration
+
+Each system has a few configurable parameters:
+
+- it may have **ordering constraints**, causing it to run before or after other systems
+  - there are several subtly different types of ordering constraints, see the next section for details
+  - e.g. `.add_system(player_controls.before(GameSets::Physics).after(CoreSets::Input))`
+- it may have one or more **run criteria** attached
+  - a system is only executed if all of its run criteria return `true`
+  - **states** are a special, more complex pattern that use run criteria to determine whether or not a system should run in a current state
+  - e.g. `.add_system(physics.run_if(resource_equals(EnablePhysics(true))))`
+- it may have one or more **labels**, allowing other systems to refer to it and enabling mass-configuration
+  - e.g. `.add_system(physics.in_set(GameSets::Physics)`
+
+This configuration is additive: adding another ordering constraint, run criteria or set does not replace the existing configuration.
+
+System configuration can be stored in a `SystemConfig` struct.
+This can be useful to reuse, compose and quickly apply complex configuration strategies, before applying these strategies to sets or individual systems.
+
+### System set configuration
+
+Each system set has its own associated `SystemConfig`, stored in the corresponding `Schedule`. When a system set is configured, its configuration is effectively passed down to every system under it. For example, ordering the set `A` to run before the set `B` specifies that all systems in `A` will run before any system in `B` does.
+
+You can add many systems to the same set at once using the `App::add_systems(systems: impl SystemIterator)` method.
+
+```rust
+#[derive(SystemLabel)]

```

Comment:
Just wanted to leave an update that "sets and labels are the same" was a miscommunication.

This RFC does not change labels. It just makes it clear that labels are *names* for things and not *tags*. If anyone thought otherwise in current Bevy, I would consider that a large misconception.

If you add multiple labels to a system, the expectation is that each label is involved in at least one dependency, so you're always positioning system(s). This RFC just makes that explicit. You position things within sets (and because sets can be "nested", we've added depth to it). I don't believe anyone was adding labels and leaving them "unconstrained" (e.g. to categorize systems by plugin).

However, such a thing has a direct analogue in this design, i.e. you can also include systems in sets that are decoupled from the ones you'd normally execute. In that case, you'd be using "set" more as an abstract group instead of something that will translate to a dependency graph.

Reply:
Copying here as requested.

> It still seems like SystemLabels are really "names of system sets".
> - Can a system have a label without being a member of the set associated with that label (I think the answer is no)?
> - Can a system set be identified by more than one label (I think the answer answer is no)? Is there a 1:1 mapping between SystemLabels and Sets (I think the answer is yes)?

So in this, a label is literally just a name. You can name a system, and you can name a system set.

You can use a label to give an instance of a system a unique name (using the `named` descriptor method). And while it might seem like labels and sets have some special connection, they don't. It's just that a system set isn't much more than a name (in this builder API).

Another way of looking at it is that the `label` method was doing two things before and we split it into `in_set` and `named`. So something can only have one name, but it can go in multiple sets.

Reply:
> Another way of looking at it is that the label method was doing two things before and we split it into in_set and named. So something can only have one name, but it can go in multiple sets.

It sounds like `system.named(X)` is the rationale behind the split. From my perspective, we shouldn't have that api and "sets" should be the one conceptual model. Even the default "system name labels" used for things like `a.after(b)`  should be a set, because we use them as cross-system-instance ordering constraints (aka ... sets). They aren't specific to an instance of the system, they apply to all systems with that name. Therefore all systems with the same name are in the same "set" (and are scheduled in the exact same way sets are). As long as we agree that sets dont need to be explicitly created (as I think we're reaching consensus on in the other thread), this seems desirable / good.

Reply:
I think you're playing too fast and loose with system sets. Your suggestion makes things too implicit.

> It sounds like `system.named(X)` is the rationale behind the split.

`system.named(X)` exists because you can call `add_system` many times with the same function and each instance needs a different name. That's it.

I wouldn't say that's *strictly* necessary, but it's much better for you when debugging if every node has a unique name.

It sounds like you're just trying to take this opportunity to get rid of the remaining "unknown label" errors by saying all names are system sets. At least, I don't see any other functional differences with your suggestion.

But I don't like the idea of promoting names to system sets for the same reason you didn't like promoting system sets to schedules. A system set is "more" than a name, so it's better for users if there's no ambiguity here.

> Therefore all systems with the same name are in the same "set".

If we took your suggestion, when a user adds a function multiple times (e.g. `apply_system_buffers`), all those instances will be grouped in a system set by default. I think that's bad. It's significantly better that you have to intentionally group systems together instead of them starting out grouped and you have to *ungroup* them.

Having a distinction between plain names and system sets is better for expressing the user's intent clearly.

Lastly, I may have said "a system set isn't much more than a name in this API", but I don't know if we can confidently say that will always be true.

Reply:
> If we took your suggestion, when a user adds a function multiple times (e.g. apply_system_buffers), all those instances will be grouped in a system set by default. I think that's bad. It's significantly better that you have to intentionally group systems together instead of them starting out grouped and you have to ungroup them.

Functionally, the current "system name labels" we use function as sets, and theres no way to make these "unique" without forcing the user to add additional information (ex: `apply_system_buffers.named(ApplyBuffers1)` and `apply_system_buffers.named(ApplyBuffers2)`).

> > system.named(X) exists because you can call add_system many times with the same function and each instance needs a different name. That's it.

Hmm if `named(X)` forces uniqueness, I'm more open to this. But I interpreted this as the `foo.after(some_system_here)` functionality, which should definitely be set-like by default (and currently behaves set-like for orderings) and should not be assumed to be unique by default (otherwise apply_system_buffers would fail on the second use in the schedule, as the names would conflict). People should feel free to add multiple copies of the same system in their schedule without needing to provide unique names.

Even if we support names, I still think set names should be `derive(SystemSet)` and system names should be `derive(SystemName)` to illustrate that they are different. But I'm also not convinced that we _need_ uniqueness. Why not just let users use "sets of size one" for cases that need unique names? Idk if the extra concept is worth it, given how similar they would be.

Reply:
I'm finding it more and more difficult to understand your mental model (and your interpretation of ours). Not sure what you mean by "this should by set-like by default" for one.

As is written, you can define a dependency (a directed edge) between any pair of nodes, where a node can be a system or a system set. It's so simple. You name stuff and then you use those names. `a.after(b)` doesn't care what kind of node a name refers to.

`system.after(name_of_set)`, `set.after(name_of_system)`, and `set.after(name_of_other_set)` all just work. Internally, we will combine that information to produce a "flattened" graph (of only system nodes and system->system edges) to compute the ultimate topsorted system order and list of dependencies for the executor to use.

> People should feel free to add multiple copies of the same system in their schedule without needing to provide a unique name.

I think people should be able to freely add dependencies without ambiguity over which nodes are involved. These two properties seem at odds. With what you suggest, it sounds like if I *don't* give systems unique names, and it's 4AM and I type something like `X.after(apply_system_buffers)`, that will apply to every single instance, added by me or a plugin. That's _way_ too implicit.

How would a user configure something to run after a specific instance of `apply_system_buffers` with what you propose?

Reply:
I'll put it another way.

If you're saying that dependencies should only exist between sets, that *can* work.

*But* auto-promoting function names into set names becomes a footgun that I find unacceptable. It would mean every auto-generated system label is an implicit system set of all nameless instances of said function. That's just straight-up bad. Because calling `X.after(some_fn)` would then mean "run after every single unnamed instance of this function," which has very high probability of inserting at least one invalid edge that triggers a "schedule unsolvable" error.

Example:

```rust
app
    .configure_set(A.before(B))
    .add_system(some_fn.in_set(A))
    .add_system(some_fn.in_set(B))
    /* ... */
    // This will create a bunch of edges in the flattened graph, and at least two of them are invalid.
    // - A can't run after something inside it.
    // - A can't run before B and run after something inside B.
    .configure_set(A.after(some_fn))
    // The kicker is the user might not even have permission to know where all these instances of `some_fn` are, since plugins could be adding them.
    /* ... */
    .run()
```

We're trying to steer people away from writing this kind of spaghetti.

This problem just doesn't exist if system instances need unique names, but I can reluctantly accept your suggestion *if* you accept that function names shouldn't automatically be set names.

Buuuut that makes the common single-instance case more verbose. People will wonder why `X.after(some_fn)` doesn't work unless they manually add the system to a set with the same name (i.e. `add_system(some_fn.in_set(some_fn))`).

I assume you don't like that either, so let's just stick to the current proposal.

Reply:
> I'm finding it hard to understand your mental model (and your interpretation of ours) here. Not sure what you mean by "this should by set-like by default" for one.
> system.after(name_of_set), set.after(name_of_system), and set.after(name_of_other_set) all just work. Internally, we will combine that information to produce a "flattened" graph (of only system nodes and system->system edges) to compute the ultimate topsorted system order and list of dependencies for the executor to use.

The point I was trying to make is that ordering using labels with the same "value" _is_ set-like, by definition. A given `a.after(b)` doesn't know how many of `b` there are.  If there is more than one thing labeled `b`, the ordering will behave like a set with the name `b`. But given your second message it sounds like you did ultimately sort out my perspective here. 

> With what you suggest, it sounds like if I don't give systems unique names, and it's 4AM and I type something like X.after(apply_system_buffers),  that will apply to every single instance in the graph, added by me or a plugin. That's way too implicit (and in this example, most likely unsolvable).

I agree that this a problem worth solving, but id like to call out that this is just a natural consequence of "labeled/set based ordering" and "systems having the same function name / label / set" (both with the current scheduler, and how i interpreted this RFC).

Looking at @alice-i-cecile's suggestion, that does help the situation, but it doesn't change the fact that the core "labeled ordering" pattern is "set-like", even with the extra validation. Labels, ultimately, refer to groups of systems in the schedule which you can order yourself relative to.

The "set api" really only adds the ability to add labels to other labels, and order labels relative to other labels. I feel like a "best of both worlds" solution exists here, where we can unify the concepts (cutting down on redundant jargon / reducing api surface), while still protecting against the things we want to (the "system function name" systems sets being used for ordering when there is more than one system in the set). Because functionally, multiple systems having the same function name _does_ put them in the same "set". I feel like adding the validation at the set level would also keep the code smaller / simpler, as we will already have the collection of systems in the set. All we're missing is a way to say that a given set/label must be unique when used for ordering.

Reply:
> It sounds like you did ultimately sort out my perspective here.

Yes, I did, and naming is no longer mandatory in the current proposal. Do we need to reach an agreement on the semantics here for you to merge the RFC?

> Because functionally, multiple systems having the same function name does put them in the same "set".

No, it doesn't. Systems do not belong to any set unless the user/plugin says they do.

> Labels, ultimately, refer to groups of systems in the schedule which you can order yourself relative to.

Cart, you added `AsSystemLabel` as a time-saver and are only now arbitrarily deciding that the coincidental "association" it creates is a set-defining property. But we should never create sets unprompted, let alone from publicly-visible labels. I've given plenty of reasons and examples to support that being too error-prone.

> I agree that this a problem worth solving.

I think you're downplaying how bad it is. I dunno how you were able to say all that stuff about how being able to retrieve a schedule for any set was too error-prone / required too much runtime validation and then do a 180° here.

In the event there are multiple (unnamed) instances of a function, the "set of all (unnamed) instances of that function" will be unusable for scheduling 9 times out of 10. I've shown how easily it breaks, and that was with the conservative assumption we only consider unnamed systems.

> cutting down on redundant jargon / reducing api surface

All this accomplishes is changing `system.named(X).in_set(Y)` into `system.in_set(X).in_set(Y)`.

And I genuinely believe that removing `named` will make things less intuitive.

An error to the tune of "You tried to create a dependency on an `AsSystemLabel` name that matches multiple system instances. A name must refer to only one instance to be used here." is easy to understand, somewhat instictively.

On the other hand, "You tried to create a dependency on an `AsSystemLabel` set. This is not allowed when said set contains more than one system. Add the systems of interest to other sets and use them." is a headscratcher.

And it's not just the weird "set of one" restriction,. We would also have to prevent users from manually adding other elements to these sets. I understand knowing "all instances of a function" has uses, but scheduling will almost never be one of them. The only things you can do that would never blow up in your face is list them out or highlight them in a visualizer. It would be simpler and more robust to have a `HashMap<TypeId, Vec<NodeId>>` for that.

> the core "labeled ordering" pattern is "set-like"

Side note, this phrase is nonsensical. What you're saying is that *systems* are single-element sets.

Reply:
> No, it doesn't. Systems do not belong to any set unless the user/plugin says they do.

I guess I haven't made myself fully clear yet, which is super fair giving that I've been saying "set" in cases where I mean "logically set-like", not "set as you've chosen to define it in this RFC". This whole time all I've been (repeatedly) trying to say is that these `LabelA.after(LabelB)` relationships behave "logically set-like" (as in, referencing an unordered group of things)  and I feel like I keep hitting a brick wall of "no thats not a set as I've defined it". For two instances of `foo()` in a schedule, they will both have the same Name/Label/Set/Whatever we choose to call `foo`. And if you say `a.after(foo)`, this is referencing everything with `foo`. This is absolutely a problem to be solved. I'm glad you identified it and chose to solve it (as it certainly wasn't on my mind when I added AsSystemLabel, nor was it a priority for me when I started reviewing this RFC).

> [In response to "I agree this is a problem worth solving"]: I think you're downplaying how bad it is. I dunno how you were able to say all that stuff about how being able to retrieve a schedule for any set was too error-prone / required too much runtime validation and then do a 180° here.

I literally said that I agree that this is a problem worth solving. Should I have said "big problem"? I'm already on your side here. I'm not doing a 180. You've identified a problem worth solving, the only question at this point, is how.

>  Cart, you added AsSystemLabel as a time-saver and are only now arbitrarily deciding that the coincidental "association" it creates is a set-defining property. But we should never form sets unprompted, let alone from publicly-visible labels. I've given plenty of reasons and examples to support that being too error-prone.

I don't think I'm deciding this arbitrarily. I agree that I added SystemLabelId/AsSystemLabel as a time-saver (without much regard for the consequences brought up here). The point I'm making is that the "logical set-like" behavior _naturally flowed from that_, given the `a.after(b)` semantics. In the context of function names as SystemLabelIds _that is a problem_. But the core problem is that we have "n > 1 instances of things with the same label, which is expressible as an ordering". Pretending that these labels don't behave like "logical unordered groups" in the context of the ordering apis we've agreed to use is _more confusing_ in my opinion. Thats just how this works.

From my perspective conceptually, the `named` approach (as proposed above) doesn't even solve this problem, and just makes it weirder. In the context of the schedule,  systems with the same "name" _are_ allowed to exist. They are "logically grouped", because they have these "shared labels" and our ordering apis allow us to reference them ambiguously with those labels. The fact that we're introducing a new concept into the mix (which behaves kind of like a set, uses the same ordering apis, uses the SystemLabel type, _and_ still allows multiple things with same name) feels more confusing, especially because it necessitates "names", "sets", _and_ "labels" conceptually.

From my perspective, embracing "set" as the _one_ concept is _just_ following the natural pattern of the apis we've already agreed to use. It vastly simplifies the mental model for users. Whenever you are labeling/ordering/naming, you are using the "set api" with the SystemSet trait. When you define a new group of systems, you derive a `SystemSet`. Which is the exact type that `in_set`, `after`, `before`, and `configure_set` deals with.

> On the other hand, "You tried to create a dependency on an AsSystemLabel set. This is not allowed when said set contains more than one system. Add the systems of interest to other sets and use them." is a headscratcher.

I think this error message is just as good as the ideal one you suggested (if not better):
"You tried to create a dependency on `fn foo()`, which refers to more than one system in the schedule. `fn foo()` is a 'system function' set, which cannot reasonably be used in ordering relationships when there is more than one instance of that system in the schedule. Consider using a different set without this restriction to express this ordering.".

 > And it's not just the weird "set of one" restriction. We would also have to prevent users from manually adding other elements to these sets.

Whats stopping people from doing this in your proposed impl?

```rust
fn some_system() {}

app
    .configure_set(some_system.after(X))
    .add_system(bar.in_set(some_system))
```

Seems like this is a problem that exists not matter what, in a world where `some_system` is a system label / convertible to one. Probably a problem worth solving, but it feels like the solution is the same for both proposals (disallow some operations for SystemLabelIds, such as users calling `in_set`).

Reply:
As promised on discord, I'll put together a branch with the changes I'm proposing for this RFC so we can talk about them concretely.

Reply:
> This whole time all I've been (repeatedly) trying to say is that these `LabelA.after(LabelB)` relationships behave "logically set-like" and I feel like I keep hitting a brick wall of "no thats not a set as I've defined it".

In my defense, the definition of set seemed like the point of this debate. I never denied that `A.after(B)` doesn't care about what's inside `A` and `B`, so it just felt like you were restating the obvious. But I get it. Whether we wanted them to be, instances of a function are linked through their `AsSystemLabel` impl, and we've said that a set is a logical group of systems. So I can understand the POV of "it's weird to pretend this group doesn't exist".

> Should I have said "big problem"?

No, and sorry for the harsh tone. For whatever reason, I thought you were trying to emphasize some use for the "logically set-like" behavior of `AsSystemLabel` labels to say that we should proudly consider them as "real" sets (as I defined them).

> But the core problem is that we have "n > 1 instances of things with the same label, which is expressible as an ordering".
> From my perspective conceptually, the `named` approach (as proposed above) doesn't even solve this problem, and just makes it weirder.
>
> Whats stopping people from [using `AsSystemLabel` to create a set] in your proposed impl?
> Seems like this is a problem that exists not matter what [and] it feels like the solution is [to] disallow some operations for some `SystemLabelId`s, such as users calling `in_set`.

In my mind, the solution to the core problem was making it an error.

I thought `in_set` and `configure_set` only accepting `impl SystemLabel` and not `impl AsSystemLabel` would prevent issues, and then `named` would be how you gave systems unique labels and we could avoid having to see them as sets (which made sense to me since you can't order around them when n > 1, re-configure them, or add anything to them). But I guess the trait bound ultimately stops nothing, since you could just do `some_system.as_label()` or whatever. And if we need to be able to provide those restrictions in general, then you have the winning argument. Like, I guess we *would* want plugins to have means to stop users from putting things in their sets.

> I think this error message is just as good if not better:
> "You tried to create a dependency on `fn foo()`, which refers to more than one system in the schedule. `fn foo()` is a 'system function' set, which cannot reasonably be used in ordering relationships when there is more than one instance of that system in the schedule. Consider using a different set without this restriction to express this ordering."

When you phrase it like this, I can get behind it.